### PR TITLE
add operator -> MiDaS computes relative inverse depth from a single i…

### PR DESCRIPTION
**Pull request by meua**
add MiDaS models for computing relative depth from a single image. from https://pytorch.org/hub/intelisl_midas_v2/

### CODE REVIEW

**Comment by meua** [comment]
> Can you provide a way to run this PR as well as details about the expected outputs?

Yes, Here are the steps to run this instance：
Steps to produce the behavior:
1. Dora start daemon: `dora up` 
2. Start a new dataflow: `dora start graphs/tutorials/webcam_single_dpt_frame.yaml --attach --hot-reload --name dpt_midas`

**Screenshots or Video**
image

**Comment by haixuanTao** [comment]
Depth Frame are very visual outputs. In the best of my capabilities, I can't picture float matrix. Would it be possible for you to export the `depth_frame` into the plot operator.

I have found an example that you can reuse: https://github.com/IntelRealSense/librealsense/blob/master/wrappers/python/examples/opencv_viewer_example.py

**Comment by meua** [comment]
> Depth Frame are very visual outputs. In the best of my capabilities, I can't picture float matrix in my mind as you screenshotted. Would it be possible for you to export the `depth_frame` into the plot operator.
> 
> I have found an example that you can reuse: https://github.com/IntelRealSense/librealsense/blob/master/wrappers/python/examples/opencv_viewer_example.py

Ok, thanks, I'll try.

**Comment by meua** [comment]
> Depth Frame are very visual outputs. In the best of my capabilities, I can't picture float matrix in my mind as you screenshotted. Would it be possible for you to export the `depth_frame` into the plot operator.
> 
> I have found an example that you can reuse: https://github.com/IntelRealSense/librealsense/blob/master/wrappers/python/examples/opencv_viewer_example.py

image

**Comment by haixuanTao** [comment]
Good, can you:
- [ ] Use `operator/plot.py` to plot the depth frame by replacing the original image
- [ ] Add a documentation page for your `midas` operator
- [ ] Clean up your code, removing unused comment
- [ ] Use black for code formatting and respect pep8

File: operators/single_dpt_op.py [hunk 1]
```diff
@@ -78,11 +82,33 @@ def on_input(
...
+                normalized_depth = 255 * (depth_output - depth_min) / (depth_max - depth_min)
+                normalized_depth *= 3
+                depth_frame = np.repeat(np.expand_dims(normalized_depth, 2), 3, axis=2) / 3
+                depth_frame = cv2.applyColorMap(np.uint8(depth_frame), cv2.COLORMAP_INFERNO)
+                original_image_bgr = np.flip(image, 2)
+                combine_image_depth_frame = np.concatenate((original_image_bgr, depth_frame), axis=1)
+                print("depth_img: ", combine_image_depth_frame)
+
+                height, width = combine_image_depth_frame.shape[:2]
+                # 缩小图像
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:101]: Please avoid chinese

Comment (meua) [reply operators/single_dpt_op.py]: ok， I will replace it

Comment (meua) [reply operators/single_dpt_op.py]: image

File: install_requirements.txt [hunk 2]
```diff
@@ -35,4 +35,13 @@ pyquaternion
...
 dora-rs==0.2.2
-tensorboard
\ No newline at end of file
+tensorboard
+
+##### MiDaS ####
+timm==0.6.12
+imutils==0.5.4
+einops==0.6.0
+torch>=1.13.0
```
Comment (haixuanTao) [inline install_requirements.txt:44]: `torch` should be already installed using: `conda install pytorch=1.11.0 torchvision=0.12.0 cudatoolkit=11.3 -c pytorch -y`
 that you can find here: https://dora.carsmos.ai/dora-drives/installation.html

There is different version for each cuda version.

Can you remove the dependencies that are installed elsewere and only install those that are strictly not already present?

File: operators/single_dpt_op.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                depth_max = depth_output.max()
+                normalized_depth = 255 * (depth_output - depth_min) / (depth_max - depth_min)
+                normalized_depth *= 3
+                depth_frame = np.repeat(np.expand_dims(normalized_depth, 2), 3, axis=2) / 3
+                depth_frame = cv2.applyColorMap(np.uint8(depth_frame), cv2.COLORMAP_INFERNO)
+                h, w = depth_frame.shape[:2]
+                depth_frame_4 = np.dstack([depth_frame, np.ones((h, w), dtype="uint8") * 255])
+                print("depth_frame: ", depth_frame_4.shape)
+                send_output("depth_frame", depth_frame_4.tobytes(), dora_input["metadata"])
+            print("frame processed")
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:93]: Remove print that are not related to any bug or unexpected behaviour.

Frame being processed is a normal behaviour

File: operators/single_dpt_op.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                # use cv2 show
+                depth_min = depth_output.min()
+                depth_max = depth_output.max()
+                normalized_depth = 255 * (depth_output - depth_min) / (depth_max - depth_min)
+                normalized_depth *= 3
+                depth_frame = np.repeat(np.expand_dims(normalized_depth, 2), 3, axis=2) / 3
+                depth_frame = cv2.applyColorMap(np.uint8(depth_frame), cv2.COLORMAP_INFERNO)
+                h, w = depth_frame.shape[:2]
+                depth_frame_4 = np.dstack([depth_frame, np.ones((h, w), dtype="uint8") * 255])
+                print("depth_frame: ", depth_frame_4.shape)
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:91]: I don't think that this is necessary

File: operators/single_dpt_op.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                depth_output = prediction.cpu().numpy()
+
+                # use cv2 show
+                depth_min = depth_output.min()
+                depth_max = depth_output.max()
+                normalized_depth = 255 * (depth_output - depth_min) / (depth_max - depth_min)
+                normalized_depth *= 3
+                depth_frame = np.repeat(np.expand_dims(normalized_depth, 2), 3, axis=2) / 3
+                depth_frame = cv2.applyColorMap(np.uint8(depth_frame), cv2.COLORMAP_INFERNO)
+                h, w = depth_frame.shape[:2]
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:89]: Can we avoid single letter variable. Can we use `height, width`

File: operators/single_dpt_op.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                prediction = self.model(input_batch)
+                prediction = torch.nn.functional.interpolate(
+                    prediction.unsqueeze(1),
+                    size=img.shape[:2],
+                    mode="bicubic",
+                    align_corners=False,
+                ).squeeze()
+                depth_output = prediction.cpu().numpy()
+
+                # use cv2 show
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:82]: This is not necessary

File: operators/single_dpt_op.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        if dora_input["id"] == "image":
+            frame = np.frombuffer(
+                dora_input["data"],
+                np.uint8,
+            ).reshape((IMAGE_HEIGHT, IMAGE_WIDTH, 4))
+
+            with torch.no_grad():
+                image = frame[:, :, :3]
+                img = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+                print("img: ", img.shape)
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:71]: This is not necessary

File: operators/single_dpt_op.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            )
+        else:
+            # Without internet
+            self.model = torch.hub.load(
+                DPT_PATH,
+                "custom",
+                path=DPT_WEIGHT_PATH,
+                source="local",
+            )
+        midas_transforms = torch.hub.load("intel-isl/MiDaS", "transforms")
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:34]: This seems to load a transform from the hub regardless of the environment variable. Is this normal? And if so, we should probably warn the user that an internet connection is happening here

File: operators/single_dpt_op.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+import cv2
+import numpy as np
+import torch
+from dora import DoraStatus
+
+IMAGE_WIDTH = 1920
+IMAGE_HEIGHT = 1080
+DEVICE = os.environ.get("PYTORCH_DEVICE") or "cpu"
+DPT_PATH = os.environ.get("DPT_PATH")
```
Comment (haixuanTao) [inline operators/single_dpt_op.py:12]: Can we avoid the abbreviation `DPT` and instead use `DEPTH`. Two letters does not break a lot of memory.

Can we also rename this `MIDAS`?

**Comment by haixuanTao** [comment]
Can we remove `single_` from the naming as we use singular and plural to differentiate between singular and plural cases?

File: graphs/tutorials/webcam_single_dpt_frame.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  - id: single_dpt
+    operator:
+      outputs:
+        - depth_frame
+      inputs:
+        image: webcam/image
+      python: ../../operators/single_dpt_op.py
+    env:
+      PYTORCH_DEVICE: "cuda"
+      DEPTH_ESTIMATION_PATH: $DORA_DEP_HOME/dependencies/MiDaS/
```
Comment (haixuanTao) [inline graphs/tutorials/webcam_single_dpt_frame.yaml:19]: It would be useful to have a guide on how to clone the Midas repo and get the weights

File: install_requirements.txt (hunk 2 above)
Comment (meua) [reply install_requirements.txt]: The corresponding dependencies have been removed
```python
- torch>=1.13.0
- torchvision>=0.14.0
```

File: operators/single_dpt_op.py (hunk 3 above)
Comment (meua) [reply operators/single_dpt_op.py:91]:
```python
print("depth_frame: ", depth_frame_4.shape)
```
This line of print code has been removed

File: operators/single_dpt_op.py (hunk 5 above)
Comment (meua) [reply operators/single_dpt_op.py:88]:
```python
-  height, width = depth_frame.shape[:2]
+ height, width = depth_frame.shape[:2]
```
single character variable replaced

File: operators/single_dpt_op.py (hunk 6 above)
Comment (meua) [reply operators/single_dpt_op.py:81]: Line break comment description removed
```python
-                            # use cv2 show
```

File: operators/single_dpt_op.py (hunk 7 above)
Comment (meua) [reply operators/single_dpt_op.py]:
```python
                print("img: ", img.shape)
```
This line of print logs has been removed

File: operators/single_dpt_op.py (hunk 9 above)
Comment (meua) [reply operators/single_dpt_op.py:12]:
```python
-  DPT_PATH = os.environ.get("DPT_PATH")
-  DPT_WEIGHT_PATH = os.environ.get("DPT_PATH")
+ MIDAS_PATH = os.environ.get("MIDAS_PATH")
+ MIDAS_WEIGHT_PATH = os.environ.get("MIDAS_WEIGHT_PATH")
```
DPT is replaced by MIDAS.

File: graphs/tutorials/webcam_single_dpt_frame.yaml (hunk 10 above)
Comment (meua) [reply graphs/tutorials/webcam_single_dpt_frame.yaml]: File names have been modified to highlight midas.

File: operators/single_dpt_op.py (hunk 8 above)
Comment (meua) [reply operators/single_dpt_op.py:34]: Yes! Thanks for pointing this out, I'll make it clear in the docs that networking is required here

File: graphs/tutorials/webcam_midas_frame.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
+nodes:
+  - id: webcam
+    operator:
+      python: ../../operators/webcam_op.py
+      inputs:
+        tick: dora/timer/millis/100
+      outputs:
+        - image
+  - id: single_dpt
```
Comment (haixuanTao) [inline graphs/tutorials/webcam_midas_frame.yaml:9]: Can we replace this with `midas` ?

File: operators/single_dpt_op.py
Comment (haixuanTao) [inline operators/single_dpt_op.py:1]: Can we remove this file?

File: operators/midas_op.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+                prediction = torch.nn.functional.interpolate(
+                    prediction.unsqueeze(1),
+                    size=img.shape[:2],
+                    mode="bicubic",
+                    align_corners=False,
+                ).squeeze()
+                depth_output = prediction.cpu().numpy()
+                depth_min = depth_output.min()
+                depth_max = depth_output.max()
+                normalized_depth = 255 * (depth_output - depth_min) / (depth_max - depth_min)
```
Comment (haixuanTao) [inline operators/midas_op.py:82]: So from what I understand, this will produce a relative depth image and not an absolute depth image. Can you check if there is a way the depth frame output can be in absolute term and if they specify a way to get the output in a measurable unit like meters?

File: docs/src/SUMMARY.md [hunk 13]
```diff
@@ -12,6 +12,7 @@
 
 - [Streaming a video](./webcam_plot.md)
 - [Yolov5](./yolov5.md)
+- [MiDaS](./webcam_midas_dpt.md)
```
Comment (haixuanTao) [inline docs/src/SUMMARY.md:15]: Can we move this into the `# Operators and Nodes`

File: docs/src/webcam_midas_dpt.md [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+     - For moderately less quality, but better speed-performance trade-off: [dpt_swin2_large_384](https://github.com/isl-org/MiDaS/releases/download/v3_1/dpt_swin2_large_384.pt)
+     - For embedded devices: [dpt_swin2_tiny_256](https://github.com/isl-org/MiDaS/releases/download/v3_1/dpt_swin2_tiny_256.pt), [dpt_levit_224](https://github.com/isl-org/MiDaS/releases/download/v3_1/dpt_levit_224.pt)
+     - For inference on Intel CPUs, OpenVINO may be used for the small legacy model: openvino_midas_v21_small [.xml](https://github.com/isl-org/MiDaS/releases/download/v3_1/openvino_midas_v21_small_256.xml), [.bin](https://github.com/isl-org/MiDaS/releases/download/v3_1/openvino_midas_v21_small_256.bin)
+
+   - MiDaS 3.0: Legacy transformer models [dpt_large_384](https://github.com/isl-org/MiDaS/releases/download/v3/dpt_large_384.pt) and [dpt_hybrid_384](https://github.com/isl-org/MiDaS/releases/download/v3/dpt_hybrid_384.pt)
+
+   - MiDaS 2.1: Legacy convolutional models [midas_v21_384](https://github.com/isl-org/MiDaS/releases/download/v2_1/midas_v21_384.pt) and [midas_v21_small_256](https://github.com/isl-org/MiDaS/releases/download/v2_1/midas_v21_small_256.pt)
```
Comment (haixuanTao) [inline docs/src/webcam_midas_dpt.md:40]: As we can see in https://github.com/dora-rs/dora-drives/issues/56 linking weights can easily be error prone. Can we just reference the Midas documentation here.

File: graphs/tutorials/webcam_midas_frame.yaml [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      outputs:
+        - depth_frame
+      inputs:
+        image: webcam/image
+      python: ../../operators/midas_op.py
+    env:
+      PYTORCH_DEVICE: "cuda"
+#      MIDAS_PATH: $DORA_DEP_HOME/dependencies/MiDaS/
+#      MIDAS_WEIGHT_PATH: $DORA_DEP_HOME/dependencies/MiDaS/weights/dpt_beit_large_512.pt
+      MODEL_TYPE: "DPT_Large"
```
Comment (haixuanTao) [inline graphs/tutorials/webcam_midas_frame.yaml:20]: Can we use the smallest model in the tutorial as some people might not have the resources to run the bigger model.

File: install_requirements.txt [hunk 16]
```diff
@@ -35,4 +35,11 @@ pyquaternion
...
-tensorboard
\ No newline at end of file
+tensorboard
+
+##### MiDaS ####
+timm==0.6.12
+imutils==0.5.4
+einops==0.6.0
+nvidia-cuda-nvrtc-cu11>=11.7.99
+nvidia-cuda-runtime-cu11>=11.7.99
```
Comment (haixuanTao) [inline install_requirements.txt:45]: I think that the cuda toolkit is installed as well with conda. Are we sure this is required?

**Review by haixuanTao — COMMENTED** [review]
We're almost there! :+1:

File: install_requirements.txt (hunk 16 above)
Comment (meua) [reply install_requirements.txt]: > I think that the cuda toolkit is installed as well with conda. Are we sure this is required?

Thank you for reminding, these dependencies have been removed and midas is running normally.

File: graphs/tutorials/webcam_midas_frame.yaml (hunk 15 above)
Comment (meua) [reply graphs/tutorials/webcam_midas_frame.yaml]: > Can we use the smallest model in this graph as some people might not have the resources to run the bigger model. This way the default is more inclusive.

The large model has been replaced with a small model.

File: docs/src/webcam_midas_dpt.md (hunk 14 above)
Comment (meua) [reply docs/src/webcam_midas_dpt.md]: > As we can see in #56 linking weights can easily be error prone. Can we just reference the Midas documentation here.

Ok, I've replaced this doc with the doc link.

File: docs/src/SUMMARY.md (hunk 13 above)
Comment (meua) [reply docs/src/SUMMARY.md]: > Can we move this into the `# Operators and Nodes`

Ok, I have modified the document index location to what you want.

File: operators/single_dpt_op.py
Comment (meua) [reply operators/single_dpt_op.py:1]: > Can we remove this file?

Ok, I will remove it right away.

File: graphs/tutorials/webcam_midas_frame.yaml (hunk 11 above)
Comment (meua) [reply graphs/tutorials/webcam_midas_frame.yaml]: > Can we replace this with `midas` ?

Ok, I have replaced the corresponding names with midas.

File: operators/midas_op.py (hunk 12 above)
Comment (meua) [reply operators/midas_op.py:84]: > So from what I understand, this will produce a relative depth image and not an absolute depth image. Can you check if there is a way the depth frame output can be in absolute term and if they specify a way to get the output in a measurable unit like meters?

Ok, what you said is a new feature, I will implement it later.

Comment (haixuanTao) [reply operators/midas_op.py:84]: I'll close this issue as its not possible at the time being.

File: operators/midas_op.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+            )
+        else:
+            # Without internet
+            self.model = torch.hub.load(
+                repo_or_dir=MIDAS_PATH,
+                model=MODEL_NAME,
+                weights=MIDAS_WEIGHT_PATH,
+                source="local",
+            )
+        midas_transforms = torch.hub.load("intel-isl/MiDaS", "transforms")
```
Comment (haixuanTao) [inline operators/midas_op.py:35]:
```suggestion
        if MIDAS_PATH is None:
            # With internet
            self.model = torch.hub.load(
                "intel-isl/MiDaS",
                MODEL_TYPE,
            )
            midas_transforms = torch.hub.load("intel-isl/MiDaS", "transforms")
        else:
            # Without internet
            self.model = torch.hub.load(
                repo_or_dir=MIDAS_PATH,
                model=MODEL_NAME,
                weights=MIDAS_WEIGHT_PATH,
                source="local",
            )
            midas_transforms = torch.hub.load(
                repo_or_dir=MIDAS_PATH, model="transforms", source="local"
            )
 
```

By adding `repo_or_dir` and `source=local` in the `torch.hub.load`, you can make midas transform offline.

**Review by haixuanTao — APPROVED** [review]
Looks good! Only one small change and we can merge this PR :100:
